### PR TITLE
IBP 3696 Move Environments to API Archictecture

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 1]
```diff
@@ -809,7 +809,7 @@ public StudyMetadata getStudyMetadataForGeolocationId(final Integer geolocationI
 				return null;
 			}
 		} catch (final HibernateException e) {
-			final String message = "Error with getStudyMetadataForGeolocationId() query from study with geoloCationId: " + geolocationId;
+			final String message = "Error with getStudyMetadataForGeolocationId() query from study with geoloCationId: " + instanceId;
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java:812]: Rename method in error message to: `getStudyMetadataForInstanceId()`

File: src/main/java/org/generationcp/middleware/dao/dms/ExperimentDao.java [hunk 2]
```diff
@@ -148,6 +148,24 @@ public ExperimentModel getExperimentByProjectIdAndLocation(final Integer project
...
+		try {
+			final Criteria criteria = this.getSession().createCriteria(this.getPersistentClass());
+			criteria.add(Restrictions.eq("geoLocation.locationId", instanceId));
+			criteria.add(Restrictions.eq("typeId", experimentType));
+			final List<ExperimentModel> list = criteria.list();
+			if (list != null && !list.isEmpty()) {
+				return list.get(0);
+			}
+		} catch (final HibernateException e) {
+			final String message = "Error at getExperimentByProjectIdAndLocation=" + instanceId
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/ExperimentDao.java:161]: Modify the error message as: Errot at `getExperimentByTypeInstanceId`

File: src/main/java/org/generationcp/middleware/dao/dms/InstanceDao.java [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class InstanceDao extends GenericDAO<ExperimentModel, Integer> {
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/InstanceDao.java:37]: Remove unused methods (these are copied from the Geolocation Refactoring branch but not actually used now):
```
getEnvironmentIds
getAllTrialEnvironments
countAllTrialEnvironments
getPropertiesForTrialEnvironments
getTrialEnvironmentDetails
getEnvironmentsForTraits
getNextInstanceNumber
getInstanceMetadata
getEnvironmentIdByStudyNameAndInstanceNumberAndProgramUUID
getExperimentIdEnvironmentMap
```

File: src/main/java/org/generationcp/middleware/dao/ims/ExperimentTransactionDAO.java [hunk 4]
```diff
@@ -97,7 +97,7 @@ public Long countPlantingTransactionsByStatus(final List<Integer> ndExperimentId
 			return criteria.list();
 		} catch (final HibernateException e) {
 			final String message =
-				"Error at getTransactionsByInstanceIds instanceIds = " + instanceIds + ", transactionType + "
+				"Error at getTransactionsByInstanceId instanceIds = " + instanceIds + ", transactionType + "
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/ims/ExperimentTransactionDAO.java:100]: The correct method name is: `getTransactionsByInstanceIds `

File: src/main/java/org/generationcp/middleware/manager/StudyDataManagerImpl.java [hunk 5]
```diff
@@ -1306,6 +1302,12 @@ public Boolean instanceExists(final Set<Integer> instanceIds) {
 		return geoLocationMap;
 	}
 
+	// TODO IBP-3305 Determine if this can be replaced with StudyDataManager#areAllInstancesExistInDataset
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/manager/StudyDataManagerImpl.java:1305]: Is this `TODO `still valid?

File: src/main/java/org/generationcp/middleware/service/impl/inventory/PlantingServiceImpl.java [hunk 6]
```diff
@@ -237,7 +237,6 @@ public void savePlanting(final Integer userId, final Integer studyId, final Inte
 			.getTransactionsByStudyId(studyId, transactionStatus, ExperimentTransactionType.PLANTING);
 	}
 
-	@Override
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/service/impl/inventory/PlantingServiceImpl.java:240]: Removing this is not part of branch work?

File: src/main/java/org/generationcp/middleware/domain/etl/MeasurementData.java [hunk 7]
```diff
@@ -35,6 +35,8 @@
 	private String dataType;
 	private Integer phenotypeId;
 	private MeasurementVariable measurementVariable;
+	private Integer instanceDataId;
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/domain/etl/MeasurementData.java:38]: As discussed, let's try to make "phenotypeId" generic in order to remove the need for dupe variable fields like `instanceDataId`

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java (hunk 1 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: Done

File: src/main/java/org/generationcp/middleware/dao/dms/ExperimentDao.java (hunk 2 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/dms/ExperimentDao.java]: Done

File: src/main/java/org/generationcp/middleware/dao/dms/InstanceDao.java (hunk 3 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/dms/InstanceDao.java:18]: Done

File: src/main/java/org/generationcp/middleware/dao/ims/ExperimentTransactionDAO.java (hunk 4 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/ims/ExperimentTransactionDAO.java:100]: Done

File: src/main/java/org/generationcp/middleware/manager/StudyDataManagerImpl.java (hunk 5 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/manager/StudyDataManagerImpl.java:1305]: Yes it's still valid

File: src/main/java/org/generationcp/middleware/service/impl/inventory/PlantingServiceImpl.java (hunk 6 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/service/impl/inventory/PlantingServiceImpl.java]: I may have accidentally removed it. I added it back.

File: src/main/java/org/generationcp/middleware/domain/etl/MeasurementData.java (hunk 7 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/domain/etl/MeasurementData.java]: Done

**Pull request merged**